server/src/routes/posts.js

```diff
@@ -123,40 +123,110 @@ router.delete('/:id', firebaseAuthMiddleware, async (req, res) => {
 
 // Upvote a post
 router.put('/:id/upvote', firebaseAuthMiddleware, async (req, res) => {
-  const currentPost = await Post.findById(req.body.id);
+  const currentPost = await Post.findById(req.params.id);
+  const currentUser = await User.findOne({ email: req.user.email });
 
-  try {
-    if (req.body.upvote_type === 'clap') {
-      const claps = currentPost.upvotes_clap;
-
-      await Post.updateOne(
-        { _id: req.body.id },
-        { upvotes_clap: claps + (req.body.upvote === true ? 1 : -1) },
-      );
+  // Check body inputs are valid
+  if (currentPost == null) {
+    res.status(400).json({ message: 'Invalid post ID' });
+    return;
+  }
+  if (currentUser == null) {
+    res.status(400).json({ message: 'Invalid user ID' });
+    return;
+  }
+  if (
+    req.body.upvote_type !== 'clap' &&
+    req.body.upvote_type !== 'laugh' &&
+    req.body.upvote_type !== 'sad'
+  ) {
+    res.status(400).json({ message: 'Invalid vote type' });
+    return;
+  }
 
-      const returnPost = await Post.findById(req.body.id);
+  try {
+    // Check if user has already voted on this post
+    const checkClap = currentUser.votes.claps.includes(req.params.id);
+    const checkLaugh = currentUser.votes.laughs.includes(req.params.id);
+    const checkSad = currentUser.votes.sads.includes(req.params.id);
+
+    let updateFlag = false;
+
+    // Current vote values
+    const claps = currentPost.upvotes_clap;
+    const laughs = currentPost.upvotes_laugh;
+    const sads = currentPost.upvotes_sad;
+
+    if (!checkClap && !checkLaugh && !checkSad) {
+      // If no current vote
+      if (req.body.upvote === true) {
+        updateFlag = true;
+      } else {
+        updateFlag = false;
+      }
+    } else if (req.body.upvote === false) {
+      // Remove current vote, dont set update flag
+      if (checkClap) {
+        await Post.updateOne({ _id: req.params.id }, { upvotes_clap: claps - 1 });
+        await User.updateOne({ _id: currentUser._id }, { $pull: { 'votes.claps': req.params.id } });
+      } else if (checkLaugh) {
+        await Post.updateOne({ _id: req.params.id }, { upvotes_laugh: laughs - 1 });
+        await User.updateOne(
+          { _id: currentUser._id },
+          { $pull: { 'votes.laughs': req.params.id } },
+        );
+      } else if (checkSad) {
+        await Post.updateOne({ _id: req.params.id }, { upvotes_sad: sads - 1 });
+        await User.updateOne({ _id: currentUser._id }, { $pull: { 'votes.sads': req.params.id } });
+      }
+    } else if (checkClap && req.body.upvote_type !== 'clap') {
```

Comment:
I was under the impression it was just laughing and sad that cancelled each other out (up /down vote) and you could stack a clap on as well

Reply:
Ah I assumed you could just have one vote max, like fb. If we can confirm the expected behaviour I'll fix it up. @mmir415 @MingChen91

Reply:
I think originally you could have multiple, but I think it's fine to have just 1 max ? @jacoballen4534 @RajanGuptaNZ  thoughts?

Reply:
Guess that makes sense